lib/using_restricted_array.rb

```diff
@@ -1,61 +1,118 @@
-require_relative 'restricted_array.rb'
+require_relative "restricted_array.rb"
 # RestrictedArray can be created using a specified size, or a random size in
 # the range of 1-20 will be chosen for you.
 # All values are integers in the range of 1-221.
 # RestrictedArray cannot be resized.
 
 # Calculates the length of the restricted array. All values are integers.
 # The restricted_array is terminated by 'nil' i.e. array[length] = nil
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: Linear or (O)n, where n is the number of items in
+# the array or the length of the input array
+# Space complexity: constant O(1)
 def length(array)
-  raise NotImplementedError
+  i = 0
+  while array[i] != nil
+    i += 1
+  end
+  return i
 end
 
 # Prints each integer values in the array
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: linear O(n), where n is the lenght of the array
+# Space complexity: constant O(1)
+
 def print_array(array)
-  raise NotImplementedError
+  len = length(array)
+  i = 0
+  while i < len
+    puts array[i]
+    i += 1
+  end
 end
 
 # For an unsorted array, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n), where n is the number of values in the array
+# Space complexity: constant O(1)
 def search(array, length, value_to_find)
-  raise NotImplementedError
+  length.times do |i|
+    return true if array[i] == value_to_find
+  end
+  return false
 end
 
 # Finds and returns the largest integer value the array
 # Assumes that the array is not sorted.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: linear O(n), where n is the size of the array
+# Space complexity: constant O(1)
 def find_largest(array, length)
-  raise NotImplementedError
+  i = 0
+  largest = 0
+  while i <= (length - 1)
+    if array[i] > largest
+      largest = array[i]
+    end
+    i += 1
+  end
+  return largest
 end
 
 # Finds and returns the smallest integer value in the array
 # Assumes that the array is not sorted.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: linear O(n), where n is the size of the array
+# Space complexity: constant O(1)
 def find_smallest(array, length)
-  raise NotImplementedError
+  i = 0
+  smallest = array[i]
+  while i <= (length - 1)
+    if array[i] < smallest
+      smallest = array[i]
+    end
+    i += 1
+  end
+  return smallest
 end
 
 # Reverses the values in the integer array in place
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: linear O(n), where n is the size of the array
+# Space complexity: constant O(1)
 def reverse(array, length)
-  raise NotImplementedError
+  i = 0
+  j = length
+  while i < j
+    a = array[i]
```

Comment:
What you have is 💯, but in case you haven't seen it yet, there's another way to swap in Ruby:
```
array[i], array[j-1] = array[j - 1], array[i]
```